File: src/argo_resource.rs
use log::{debug, error, warn};
use regex::Regex;
use serde_yaml::Mapping;
use std::error::Error;

use crate::{parsing::K8sResource, selector::Operator, Selector};

const ANNOTATION_WATCH_PATTERN: &str = "argocd-diff-preview/watch-pattern";
const ANNOTATION_IGNORE: &str = "argocd-diff-preview/ignore";

#[derive(PartialEq, Clone)]
pub enum ApplicationKind {
    Application,
    ApplicationSet,
}

#[derive(Clone)]
pub struct ArgoResource {
    pub yaml: serde_yaml::Value,
    pub kind: ApplicationKind,
    pub name: String,
    pub namespace: String,
    // Where the resource was found
    pub file_name: String,
}

impl ArgoResource {
    pub fn as_string(&self) -> Result<String, Box<dyn Error>> {
        Ok(serde_yaml::to_string(&self.yaml)?)
    }
}

impl PartialEq for ArgoResource {
    fn eq(&self, other: &Self) -> bool {
        self.yaml == other.yaml
    }
}

impl ArgoResource {
    pub fn set_project_to_default(mut self) -> Result<ArgoResource, Box<dyn Error>> {
        let spec = match self.kind {
            ApplicationKind::Application => self.yaml["spec"].as_mapping_mut(),
            ApplicationKind::ApplicationSet => {
                self.yaml["spec"]["template"]["spec"].as_mapping_mut()
            }
        };

        match spec {
            None => Err(format!("No 'spec' key found in Application: {}", self.name).into()),
            Some(spec) => {
                spec["project"] = serde_yaml::Value::String("default".to_string());
                Ok(self)
            }
        }
    }

    pub fn point_destination_to_in_cluster(mut self) -> Result<ArgoResource, Box<dyn Error>> {
        let spec = match self.kind {
            ApplicationKind::Application => self.yaml["spec"].as_mapping_mut(),
            ApplicationKind::ApplicationSet => {
                self.yaml["spec"]["template"]["spec"].as_mapping_mut()
            }
        };

        match spec {
            None => Err(format!("No 'spec' key found in Application: {}", self.name).into()),
            Some(spec) if spec.contains_key("destination") => {
                spec["destination"]["name"] = serde_yaml::Value::String("in-cluster".to_string());
                spec["destination"]
                    .as_mapping_mut()
                    .map(|a| a.remove("server"));
                Ok(self)
            }
            Some(_) => Err(format!(
                "No 'spec.destination' key found in Application: {}",
                self.name
            )
            .into()),
        }
    }

    pub fn remove_sync_policy(mut self) -> ArgoResource {
        let spec = match self.kind {
            ApplicationKind::Application => self.yaml["spec"].as_mapping_mut(),
            ApplicationKind::ApplicationSet => {
                self.yaml["spec"]["template"]["spec"].as_mapping_mut()
            }
        };
        match spec {
            Some(spec) => {
                spec.remove("syncPolicy");
            }
            None => debug!(
                "Can't remove 'syncPolicy' because 'spec' key not found in file: {}",
                self.file_name
            ),
        }
        self
    }

    pub fn redirect_sources(
        mut self,
        repo: &str,
        branch: &str,
    ) -> Result<ArgoResource, Box<dyn Error>> {
        let spec = match self.kind {
            ApplicationKind::Application => self.yaml["spec"].as_mapping_mut(),
            ApplicationKind::ApplicationSet => {
                self.yaml["spec"]["template"]["spec"].as_mapping_mut()
            }
        };

        match spec {
            None => Err(format!("No 'spec' key found in Application: {}", self.name).into()),
            Some(spec) if spec.contains_key("source") => {
                if spec["source"]["chart"].as_str().is_some() {
                    return Ok(self);
                }
                match spec["source"]["repoURL"].as_str() {
                    Some(url) if url.to_lowercase().contains(&repo.to_lowercase()) => {
                        spec["source"]["targetRevision"] =
                            serde_yaml::Value::String(branch.to_string());
                    }
                    _ => debug!(
                        "Found no 'repoURL' under spec.source in file: {}",
                        self.file_name
                    ),
                }
                Ok(self)
            }
            Some(spec) if spec.contains_key("sources") => {
                if let Some(sources) = spec["sources"].as_sequence_mut() {
                    for source in sources {
                        if source["chart"].as_str().is_some() {
                            continue;
                        }
                        match source["repoURL"].as_str() {
                            Some(url) if url.to_lowercase().contains(&repo.to_lowercase()) => {
                                source["targetRevision"] =
                                    serde_yaml::Value::String(branch.to_string());
                            }
                            _ => debug!(
                                "Found no 'repoURL' under spec.sources[] in file: {}",
                                self.file_name
                            ),
                        }
                    }
                }
                Ok(self)
            }
            Some(_) => Err(format!(
                "No 'spec.source' or 'spec.sources' key found in Application: {}",
                self.name
            )
            .into()),
        }
    }

    pub fn redirect_generators(
        mut self,
        repo: &str,
        branch: &str,
    ) -> Result<ArgoResource, Box<dyn Error>> {
        if self.kind != ApplicationKind::ApplicationSet {
            return Ok(self);
        }

        let spec = self.yaml["spec"].as_mapping_mut();

        match spec {
            None => Err(format!("No 'spec' key found in ApplicationSet: {}", self.name).into()),
            Some(spec) => {
                if spec.contains_key("generators") {
                    if redirect_git_generator(spec, repo, branch) {
                        debug!(
                            "Patched git generators in ApplicationSet: {} in file: {}",
                            self.name, self.file_name
                        );
                    }
                    if let Some(g) = spec["generators"].as_sequence_mut() {
                        for generator in g {
                            if let Some(i) = generator["matrix"].as_mapping_mut() {
                                if redirect_git_generator(i, repo, branch) {
                                    debug!(
                                        "Patched git generators in matrix generators in ApplicationSet: {} in file: {}",
                                        self.name,
                                        &self.file_name
                                    );
                                }
                            }
                        }
                    }
                }
                Ok(self)
            }
        }
    }

    pub fn from_k8s_resource(k8s_resource: K8sResource) -> Option<ArgoResource> {
        let kind = k8s_resource.yaml["kind"]
            .as_str()
            .and_then(|kind| match kind {
                "Application" => Some(ApplicationKind::Application),
                "ApplicationSet" => Some(ApplicationKind::ApplicationSet),
                _ => None,
            })?;

        let namespace = k8s_resource.yaml["metadata"]["namespace"]
            .as_str()
            .unwrap_or("default");

        match k8s_resource.yaml["metadata"]["name"].as_str() {
            Some(name) => Some(ArgoResource {
                kind,
                file_name: k8s_resource.file_name,
                name: name.to_string(),
                namespace: namespace.to_string(),
                yaml: k8s_resource.yaml,
            }),
            _ => None,
        }
    }

    pub fn filter(
        self,
        selector: &Option<Vec<Selector>>,
        files_changed: &Option<Vec<String>>,
        ignore_invalid_watch_pattern: bool,
    ) -> Option<ArgoResource> {
        // check if the application should be ignored
        if self.yaml["metadata"]["annotations"][ANNOTATION_IGNORE].as_str() == Some("true") {
            debug!(
                "Ignoring application {:?} due to '{}=true' in file: {}",
                self.name, ANNOTATION_IGNORE, self.file_name
            );
            return None;
        }

        // loop over labels and check if the selector matches
        if let Some(selector) = selector {
            let labels: Vec<(&str, &str)> = {
                match self.yaml["metadata"]["labels"].as_mapping() {
                    Some(m) => m
                        .iter()
                        .flat_map(|(k, v)| Some((k.as_str()?, v.as_str()?)))
                        .collect(),
                    None => Vec::new(),
                }
            };
            let selected = selector.iter().all(|l| match l.operator {
                Operator::Eq => labels.iter().any(|(k, v)| k == &l.key && v == &l.value),
                Operator::Ne => labels.iter().all(|(k, v)| k != &l.key || v != &l.value),
            });
            if !selected {
                debug!(
                    "Ignoring application {:?} due to label selector mismatch in file: {}",
                    self.name, self.file_name
                );
                return None;
            } else {
                debug!(
                    "Selected application {:?} due to label selector match in file: {}",
                    self.name, self.file_name
                );
            }
        }

        // Check watch pattern annotation
        let pattern_annotation =
            self.yaml["metadata"]["annotations"][ANNOTATION_WATCH_PATTERN].as_str();
        let list_of_regex_results = pattern_annotation.map(|s| {
            s.split(',')
                .map(|s| Regex::new(s.trim()))
                .collect::<Vec<Result<Regex, regex::Error>>>()
        });

        // Return early if a regex pattern is invalid
        if let Some(pattern_vec) = &list_of_regex_results {
            if let Some(p) = pattern_vec.iter().filter_map(|r| r.as_ref().err()).next() {
                if ignore_invalid_watch_pattern {
                    warn!("🚨 Ignoring application {:?} due to invalid regex pattern in '{}' ({}) - Error: {}",
                        self.name,
                        pattern_annotation.unwrap_or("unknown"),
                        self.file_name,
                        p);
                } else {
                    error!(
                        "🚨 Application {:?} has an invalid regex pattern in '{}' ({}) - Error: {}",
                        self.name,
                        pattern_annotation.unwrap_or("unknown"),
                        self.file_name,
                        p
                    );
                    panic!("Invalid regex pattern in annotation");
                }
            }
        }

        let patterns: Option<Vec<Regex>> =
            list_of_regex_results.map(|v| v.into_iter().flat_map(|r| r.ok()).collect());

        match (files_changed, patterns) {
            (None, _) => {}
            // Check if the application changed.
            (Some(files_changed), _) if files_changed.contains(&self.file_name) => {
                debug!(
                    "Selected application {:?} due to file change in file: {}",
                    self.name, self.file_name
                );
            }
            // Check if the application changed and the regex pattern matches.
            (Some(files_changed), Some(pattern))
                if files_changed
                    .iter()
                    .any(|f| pattern.iter().any(|r| r.is_match(f))) =>
            {
                debug!(
                    "Selected application {:?} due to regex pattern '{}' matching changed files",
                    self.name,
                    pattern
                        .iter()
                        .map(|r| r.as_str())
                        .collect::<Vec<&str>>()
                        .join(", "),
                );
            }
            (_, Some(pattern)) => {
                debug!(
                    "Ignoring application {:?} due to regex pattern '{}' not matching changed files",
                    self.name,
                    pattern.iter().map(|r| r.as_str()).collect::<Vec<&str>>().join(", "),
                );
                return None;
            }
            (_, None) => {
                debug!(
                    "Ignoring application {:?} due to missing '{}' annotation ({})",
                    self.name, &ANNOTATION_WATCH_PATTERN, self.file_name
                );
                return None;
            }
        }

        Some(self)
    }
}

// Returns true if the generators were patched
fn redirect_git_generator(v: &mut Mapping, repo: &str, branch: &str) -> bool {
    let mut patched = false;
    if v.contains_key("generators") {
        if let Some(i) = v["generators"].as_sequence_mut() {
            for generator in i {
                if let Some(git) = generator["git"].as_mapping_mut() {
                    if git.contains_key("repoURL") {
                        match git["repoURL"].as_str() {
                            Some(url) if url.to_lowercase().contains(&repo.to_lowercase()) => {
                                git["revision"] = serde_yaml::Value::String(branch.to_string());
                                debug!("Redirected 'repoURL' in git generator",);
                                patched = true;
                            }
                            Some(_url) => {
                                debug!("Found no matching 'repoURL' in git generator")
                            }
                            _ => (),
                        }
                    }
                }
            }
        }
    }
    patched
}

File: pkg/types/argo_resource.go
package types

import (
	"fmt"
	"regexp"
	"strings"

	yamlutil "github.com/argocd-diff-preview/argocd-diff-preview/pkg/yaml"
	"github.com/rs/zerolog/log"

	"gopkg.in/yaml.v3"
)

const (
	AnnotationWatchPattern = "argocd-diff-preview/watch-pattern"
	AnnotationIgnore       = "argocd-diff-preview/ignore"
)

// ApplicationKind represents the type of Argo CD application
type ApplicationKind int

const (
	Application ApplicationKind = iota
	ApplicationSet
)

// String returns the string representation of ApplicationKind
func (k ApplicationKind) String() string {
	switch k {
	case Application:
		return "Application"
	case ApplicationSet:
		return "ApplicationSet"
	default:
		return "Unknown"
	}
}

// ArgoResource represents an Argo CD Application or ApplicationSet
type ArgoResource struct {
	Yaml     *yaml.Node
	Kind     ApplicationKind
	Name     string
	FileName string
}

// AsString returns the YAML representation of the resource
func (a *ArgoResource) AsString() (string, error) {
	bytes, err := yaml.Marshal(a.Yaml)
	if err != nil {
		return "", fmt.Errorf("failed to marshal yaml: %w", err)
	}
	return string(bytes), nil
}

// SetNamespace sets the namespace of the resource
func (a *ArgoResource) SetNamespace(namespace string) error {
	yamlutil.SetYamlValue(a.Yaml, []string{"metadata", "namespace"}, namespace)
	return nil
}

// SetProjectToDefault sets the project to "default"
func (a *ArgoResource) SetProjectToDefault() error {
	spec := a.getAppSpec()
	if spec == nil {
		log.Debug().Msgf("no 'spec' key found in Application: %s", a.Name)
	}

	project := yamlutil.GetYamlValue(spec, []string{"project"})
	if project == nil {
		log.Debug().Msgf("no 'spec.project' key found in Application: %s (file: %s)",
			a.Name, a.FileName)
	}

	yamlutil.SetYamlValue(spec, []string{"project"}, "default")
	return nil
}

// PointDestinationToInCluster updates the destination to point to the in-cluster service
func (a *ArgoResource) PointDestinationToInCluster() error {
	spec := a.getAppSpec()
	if spec == nil {
		log.Debug().Msgf("no 'spec' key found in Application: %s", a.Name)
	}

	dest := yamlutil.GetYamlValue(spec, []string{"destination"})
	if dest == nil {
		log.Debug().Msgf("no 'spec.destination' key found in Application: %s (file: %s)",
			a.Name, a.FileName)
	}

	yamlutil.SetYamlValue(dest, []string{"name"}, "in-cluster")
	yamlutil.RemoveYamlValue(dest, []string{"server"})
	return nil
}

// RemoveSyncPolicy removes the syncPolicy from the resource
func (a *ArgoResource) RemoveSyncPolicy() error {
	if a.Yaml == nil {
		log.Warn().Str("patchType", "removeSyncPolicy").Str("file", a.FileName).Msgf("Can't remove 'syncPolicy' because YAML is nil")
		return nil
	}

	if a.Yaml.Content == nil {
		log.Warn().Str("patchType", "removeSyncPolicy").Str("file", a.FileName).Msgf("Can't remove 'syncPolicy' because YAML content is nil")
		return nil
	}

	spec := yamlutil.GetYamlValue(a.Yaml, []string{"spec"})
	if spec == nil {
		log.Warn().Str("patchType", "removeSyncPolicy").Str("file", a.FileName).Msgf("Can't remove 'syncPolicy' because 'spec' key not found")
		return nil
	}

	yamlutil.RemoveYamlValue(spec, []string{"syncPolicy"})
	return nil
}

// RedirectSources updates the source/sources targetRevision to point to the specified branch
func (a *ArgoResource) RedirectSources(repo, branch string, redirectRevisions []string) error {
	spec := a.getAppSpec()
	if spec == nil {
		log.Warn().Str("patchType", "redirectSources").Str("file", a.FileName).Msgf("⚠️ No 'spec' key found in Application: %s", a.Name)
	}

	// Handle single source
	source := yamlutil.GetYamlValue(spec, []string{"source"})
	if source != nil {
		if err := a.redirectSource(source, repo, branch, redirectRevisions); err != nil {
			return err
		}
		return nil
	}

	// Handle multiple sources
	sources := yamlutil.GetYamlValue(spec, []string{"sources"})
	if sources != nil {
		for _, src := range sources.Content {
			if err := a.redirectSource(src, repo, branch, redirectRevisions); err != nil {
				return err
			}
		}
		return nil
	}

	log.Debug().Str("patchType", "redirectSources").Str("file", a.FileName).Msgf("no 'spec.source' or 'spec.sources' key found in Application: %s",
		a.Name)

	return nil
}

// RedirectGenerators updates the git generator targetRevision to point to the specified branch
func (a *ArgoResource) RedirectGenerators(repo, branch string, redirectRevisions []string) error {
	// Only process ApplicationSets
	if a.Kind != ApplicationSet {
		return nil
	}

	spec := a.getRootSpec()
	if spec == nil {
		log.Warn().Str("patchType", "redirectGenerators").Str("file", a.FileName).Msgf("⚠️ No 'spec' key found in ApplicationSet: %s", a.Name)
		return nil
	}

	generators := yamlutil.GetYamlValue(spec, []string{"generators"})
	if generators == nil {
		log.Debug().Str("patchType", "redirectGenerators").Str("file", a.FileName).Str("branch", branch).Msgf("no 'spec.generators' key found in ApplicationSet: %s", a.Name)
		return nil
	}

	log.Debug().Str("patchType", "redirectGenerators").Str("file", a.FileName).Str("branch", branch).Msgf("found %d generators in ApplicationSet: %s", len(generators.Content), a.Name)

	// Process each generator
	for index, generator := range generators.Content {
		if generator.Kind != yaml.MappingNode {
			continue
		}

		// Look for git generator
		gitGen := yamlutil.GetYamlValue(generator, []string{"git"})
		if gitGen == nil {
			log.Debug().Str("patchType", "redirectGenerators").Str("file", a.FileName).Str("branch", branch).Msgf("no 'spec.generators[%d].git' key found in ApplicationSet: %s", index, a.Name)
			continue
		}

		// Check repoURL
		repoURL := yamlutil.GetYamlValue(gitGen, []string{"repoURL"})
		if repoURL == nil || !containsIgnoreCase(repoURL.Value, repo) {
			log.Debug().Str("patchType", "redirectGenerators").Str("file", a.FileName).Str("branch", branch).Msgf("no 'spec.generators[%d].git.repoURL' key found in ApplicationSet: %s", index, a.Name)
			continue
		}

		// Check targetRevision
		revision := yamlutil.GetYamlValue(gitGen, []string{"revision"})
		if revision == nil {
			log.Debug().Str("patchType", "redirectGenerators").Str("file", a.FileName).Str("branch", branch).Msgf("no 'spec.generators[%d].git.revision' key found in ApplicationSet: %s", index, a.Name)
			continue
		}

		// Check if we should redirect this revision
		shouldRedirect := len(redirectRevisions) == 0 || contains(redirectRevisions, revision.Value)
		if shouldRedirect {
			yamlutil.SetYamlValue(gitGen, []string{"revision"}, branch)
			log.Debug().Str("patchType", "redirectGenerators").Str("file", a.FileName).Str("branch", branch).Msgf(
				"Patched git generators in ApplicationSet: %s",
				a.Name,
			)
		}
	}

	return nil
}

// TODO: should ignoreInvalidWatchPattern throw error or return nil?

// Filter checks if the application matches the given selectors and watches the given files
func (a *ArgoResource) Filter(
	selectors []Selector,
	filesChanged []string,
	ignoreInvalidWatchPattern bool,
) *ArgoResource {
	// Check selectors
	if len(selectors) > 0 {
		metadata := yamlutil.GetYamlValue(a.Yaml, []string{"metadata"})
		if metadata == nil {
			return nil
		}

		labels := yamlutil.GetYamlValue(metadata, []string{"labels"})
		if labels == nil {
			return nil
		}

		for _, selector := range selectors {
			labelValue := yamlutil.GetYamlValue(labels, []string{selector.Key})
			if labelValue == nil {
				return nil
			}

			matches := labelValue.Value == selector.Value
			if (selector.Operator == Eq && !matches) || (selector.Operator == Ne && matches) {
				return nil
			}
		}
	}

	// Check files changed
	if len(filesChanged) > 0 {

		log.Debug().Str("patchType", "filter").Str("file", a.FileName).Msgf("checking files changed: %v", filesChanged)

		metadata := yamlutil.GetYamlValue(a.Yaml, []string{"metadata"})
		if metadata == nil {
			return nil
		}

		annotations := yamlutil.GetYamlValue(metadata, []string{"annotations"})
		if annotations == nil {
			return nil
		}

		watchPattern := yamlutil.GetYamlValue(annotations, []string{AnnotationWatchPattern})
		if watchPattern == nil {
			log.Debug().Str("patchType", "filter").Str("file", a.FileName).Msgf("no watch pattern annotation found. Skipping")
			return nil
		}

		patternList := strings.TrimSpace(watchPattern.Value)
		if patternList == "" {
			log.Debug().Str("patchType", "filter").Str("file", a.FileName).Msgf("no watch pattern value found. Skipping")
			return nil
		}

		patterns := strings.Split(patternList, ",")

		for _, pattern := range patterns {
			pattern = strings.TrimSpace(pattern)

			log.Debug().Str("patchType", "filter").Str("file", a.FileName).Msgf("checking watch pattern: %s", pattern)

			if pattern == "" {
				log.Debug().Str("patchType", "filter").Str("file", a.FileName).Msgf("empty watch pattern found. Continuing")
				continue
			}

			regex, err := regexp.Compile(pattern)
			if err != nil {
				if !ignoreInvalidWatchPattern {
					log.Warn().Msgf("⚠️ Invalid watch pattern '%s' in file: %s", pattern, a.FileName)
					return nil
				}
				log.Warn().Msgf("⚠️ Ignoring invalid watch pattern '%s' in file: %s", pattern, a.FileName)
				return a
			}

			log.Debug().Str("patchType", "filter").Str("file", a.FileName).Msgf("watch pattern '%s' is valid. Checking files changed", pattern)

			for _, file := range filesChanged {
				if regex.MatchString(file) {
					log.Debug().Str("patchType", "filter").Str("file", a.FileName).Msgf("file '%s' matches watch pattern. Returning application", file)
					return a
				}
			}
		}

		log.Debug().Str("patchType", "filter").Str("file", a.FileName).Msgf("no files changed match watch pattern. Skipping")
		return nil
	}

	return a
}

// Helper functions

func (a *ArgoResource) getAppSpec() *yaml.Node {
	switch a.Kind {
	case Application:
		return yamlutil.GetYamlValue(a.Yaml, []string{"spec"})
	case ApplicationSet:
		return yamlutil.GetYamlValue(a.Yaml, []string{"spec", "template", "spec"})
	default:
		return nil
	}
}

func (a *ArgoResource) getRootSpec() *yaml.Node {
	return yamlutil.GetYamlValue(a.Yaml, []string{"spec"})
}

func (a *ArgoResource) redirectSource(source *yaml.Node, repo, branch string, redirectRevisions []string) error {

	if yamlutil.GetYamlValue(source, []string{"chart"}) != nil {
		log.Debug().Str("patchType", "redirectSource").Str("file", a.FileName).Msg("Found helm chart")
		return nil
	}

	repoURL := yamlutil.GetYamlValue(source, []string{"repoURL"})
	if repoURL == nil || !containsIgnoreCase(repoURL.Value, repo) {
		log.Debug().Str("patchType", "redirectSource").Str("file", a.FileName).Msg("Found no 'repoURL' under spec.source")
		return nil
	}

	targetRev := yamlutil.GetYamlValue(source, []string{"targetRevision"})
	if targetRev == nil {
		log.Debug().Str("patchType", "redirectSource").Str("file", a.FileName).Msg("Found no 'targetRevision' under spec.source")
		targetRev = &yaml.Node{Value: "HEAD"}
		yamlutil.SetYamlValue(source, []string{"targetRevision"}, "HEAD")
	}

	shouldRedirect := len(redirectRevisions) == 0 ||
		contains(redirectRevisions, targetRev.Value)

	if shouldRedirect {
		log.Debug().Str("patchType", "redirectSource").Str("file", a.FileName).Msgf("Redirecting targetRevision from %s to %s", targetRev.Value, branch)
		yamlutil.SetYamlValue(source, []string{"targetRevision"}, branch)
	}

	return nil
}

func containsIgnoreCase(s, substr string) bool {
	return strings.Contains(strings.ToLower(s), strings.ToLower(substr))
}

func contains(slice []string, str string) bool {
	for _, s := range slice {
		if s == str {
			return true
		}
	}
	return false
}

// FromK8sResource creates an ArgoResource from a K8sResource
func FromK8sResource(resource *K8sResource) *ArgoResource {
	// Get the kind

	if resource.Yaml.Content == nil {
		log.Debug().Str("file", resource.FileName).Msg("No content found in file")
		return nil
	}

	kind := yamlutil.GetYamlValue(resource.Yaml.Content[0], []string{"kind"})
	if kind == nil {
		log.Debug().Str("file", resource.FileName).Msg("No 'kind' field found in file")
		return nil
	}

	// Check if it's an Argo CD resource
	var appKind ApplicationKind
	switch kind.Value {
	case "Application":
		appKind = Application
	case "ApplicationSet":
		appKind = ApplicationSet
	default:
		return nil
	}

	name := yamlutil.GetYamlValue(resource.Yaml.Content[0], []string{"metadata", "name"})
	if name == nil {
		log.Debug().Str("file", resource.FileName).Msg("No 'metadata.name' field found in file")
		return nil
	}

	return &ArgoResource{
		Yaml:     resource.Yaml.Content[0],
		Kind:     appKind,
		Name:     name.Value,
		FileName: resource.FileName,
	}
}

func ApplicationsToString(apps []ArgoResource) string {
	var yamlStrings []string
	for _, app := range apps {
		yamlStr, err := app.AsString()
		if err != nil {
			log.Debug().Err(err).Str("file", app.FileName).Msgf("Failed to convert app %s to YAML", app.Name)
			continue
		}
		// add a comment with the name of the file
		yamlStr = fmt.Sprintf("# File: %s\n%s", app.FileName, yamlStr)

		yamlStrings = append(yamlStrings, yamlStr)
	}
	return strings.Join(yamlStrings, "---\n")
}
